Fix more color/ansi issues

eatkins opened the pull request:
This PR fixes #5995. It also does some additional refactoring to ensure that color can be enabled independently from general ansi codes. I tested that color codes _and_ newlines were displayed on travis ci in github.com/swoval/swoval if I published a local version of sbt with these changes in the ci build. Using 1.4.1 reproduced the issue in #5995.  I also tested manually that the sbt interactive shell worked ok with different variations of -Dsbt.color, -Dsbt.log.format and -Dsbt.supershell.

eatkins committed:
Improve log formatting logic for ci use cases

It is currently the case that all tagged log messages going through a
ConsoleAppender have color codes added and then subsequently stripped
out. This didn't work well in combination with -Dsbt.color=true and
-Dsbt.ci=true because the -Dsbt.color=true was causing
ConsoleAppender.formatEnabledInEnv to be set to true which caused ansi
codes to be enabled. Travis CI supports color codes but not other ansi
escape sequences (this is also often true of embedded shells like in
intellij or jedit). This commit reworks the ConsoleAppender so that we
only add colors if useFormat is enable dand only add
ClearScreenAfterCursor if ansi codes are supported. It also reworks the
ConsoleAppender.write method so that if useFormat is true but
ansiCodesSupported is false that we only remove the non color related
ansi escape sequences so that colors remain.

eatkins committed:
Manage ansi codes and color codes separately

The ConsoleAppender formatEnabledInEnv field was being used both as an
indicator that ansi codes were supported and that color codes are
enabled. There are cases in which general ansi codes are not supported
but color codes are and these use cases need to be handled separately.
To make things more explicit, this commit adds isColorEnabled and
isAnsiSupported to the Terminal companion object so that we can be more
specific about what the requirements are (general ansi escape codes or
just colors). There are a few cases in ConsoleAppender itself where
formatEnabledInEnv was used to set flags for both color and ansi codes.
When that is the case, we use Terminal.isAnsiSupported because when that
is true, colors should at least work but there are terminals that
support color but not general ansi escape codes.

eatkins committed:
Enable ansi codes if supershell is enabled

Running sbt with -Dsbt.supershell=true and -Dsbt.color=true doesn't work
well because the isAnsiSupported flag is set to false when
-Dsbt.color=false. This causes the processing that ensures that
supershell lines do not get interleaved with normal log lines to be
skipped. To fix this, we can enable ansi codes when supershell is
explicitly enabled.

eatkins committed:
Don't ever use jline 3 dumb terminal

When using the DumbTerminal, no input is successfully read. I must not
have actually tested this when I added it.

eatkins committed:
Only disable virtual io automatically in ci

Disabling virtual io prevents the thin client from working. It is
possible that a user may manually start a server with -Dsbt.color=false
but still want to connect a thin client.

eed3si9n commented:
>  Travis CI supports color codes but not other ansi escape sequences (this is also often true of embedded shells like in intellij or jedit)

In general, could we also try to address https://github.com/sbt/sbt/issues/5269?
We regressed to not display color in CIs (https://github.com/sbt/util/pull/184/commits/53c9b848580376544543b0732848005b37d97fd7), and the users probably shouldn't have to deal with various combination to get this working for common CIs.

eatkins committed:
Enable color by default in ci builds

This channge should turn color on by default in ci builds (see
https://github.com/sbt/sbt/issues/5269). It can still be disabled with
-Dsbt.color=false.

eed3si9n reviewed (dismissed):
Thanks!

eatkins commented:
I verified that colors now appear on travis without having to set any system properties.

eatkins commented:
FTR, the colors have reappeared in sbt's ci in the build that dogfoods sbt with a publishlocal: https://travis-ci.org/github/sbt/sbt/jobs/737809980
